### ISSUE TITLE
Windows implementation or linux client (juce6) missing

### ISSUE THREAD

**Issue by elli-one** [apohl79/audiogridder#2]
Interesting project.  I miss Windows (with Juce I can see no real problem)  or a linux implementation.
The Juce socket implementation is good usable. For the audio stream maybe RTP is a better choice. 

For years I have implemented the same (but with plugin UI and more complex - see https://www.dawconnect.com/).
elli

**Comment by apohl79** [comment]
The code is mostly using juce where possible. The com code is should be working cross platform already. What has to be ported is the screen capturing code and the keyboard/mouse event handling. If you have the possibility to look at this for either windows or linux that would be awesome!

Regarding the audio streaming, RTP would add another layer of complexity and latency and would not help. AG is directly streaming sample buffers as they get provided by a DAW.

**Comment by apohl79** [comment]
Looking at RTP might indeed be interesting. Thx for the hint.

**Comment by elli-one** [comment]
Re: [apohl79/audiogridder] Windows implementation or linux client (juce6) missing (#2)


JRTP https://github.com/j0r1/JRTPLIB  is a good start point. Simple to integrate.
I think you know, UDP can be a little bit faster than TCP. In my experience earlier
I have had normally a answer from remote plugin in some mili seconds. 





Looking at RTP might indeed be interesting. Thx for the hint.

**Comment by apohl79** [comment]
> JRTP https://github.com/j0r1/JRTPLIB  is a good start point. Simple to integrate.

Yeah i saw it. Discontinued unfortunately. Thx!

> I think you know, UDP can be a little bit faster than TCP. In my experience earlier
> I have had normally a answer from remote plugin in some mili seconds. 

It  should probably improve the reliability especially on higher latency connections like wireless.

**Comment by elli-one** [comment]
Re: [apohl79/audiogridder] Windows implementation or linux client (juce6) missing (#2)


After some inserts  #if defined (JUCE_MAC) ... 
 - some changes for config file names 
 - screen capture changed to createComponentSnapshot
 - no key handling yet (later)
The server is started for windows. Only started. Tomorrow I try it for client.







JRTP https://github.com/j0r1/JRTPLIB  is a good start point. Simple to integrate.
Yeah i saw it. Discontinued unfortunately. Thx!
I think you know, UDP can be a little bit faster than TCP. In my experience earlier
I have had normally a answer from remote plugin in some mili seconds.
It should probably improve the reliability especially on higher latency connections like wireless.

**Comment by rcrath** [comment]
I'm happy to be a tester for Windows and Linux. Don't have a working Mac. Can usually manage to build stuff from source successfully unless the build is too esoteric. Familiar with git. Want to get my feet wet with juce as well ([before ilok puts a dongle on it](https://www.pro-tools-expert.com/production-expert-1/2020/4/19/plugin-code-developer-juce-acquired-by-pace-the-company-behind-ilok)). Will see how far I get on the latest source.

**Comment by bazzdream** [comment]
Will it be possible to use a powerful Windows PC as a server and a Mac as client?

**Comment by apohl79** [comment]
Yep that should be possible.

**Comment by prene1** [comment]
> Interesting project. I miss Windows (with Juce I can see no real problem) or a linux implementation.
> The Juce socket implementation is good usable. For the audio stream maybe RTP is a better choice.
> 
> For years I have implemented the same (but with plugin UI and more complex - see https://www.dawconnect.com/).
> elli

Did they EVER get this working ? Because I was so interested in using this and NEVER got it working. That how I ended up with Thousands of dollars of UAD stuff lol.

**Comment by prene1** [comment]
> Yep that should be possible.

I would love to help in beta testing on the windows side. Thanks.

**Comment by KottV** [comment]
Hi,

I've succesfuly build Linux client, just add
#include <fcntl.h> into Client.cpp

Server falls to link due unimplemented screen capture method (I guess):

```Linking AudioGridderServer - App
/usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: build/intermediate/Release/App_e5d18802.o: in function `non-virtual thunk to e47::App::ProcessorWindow::timerCallback()':
App.cpp:(.text._ZN3e473App15ProcessorWindow13timerCallbackEv[_ZN3e473App15ProcessorWindow13timerCallbackEv]+0x38): undefined reference to `e47::captureScreen(juce::Rectangle<int>)'
/usr/lib64/gcc/x86_64-suse-linux/9/../../../../x86_64-suse-linux/bin/ld: build/intermediate/Release/App_e5d18802.o: in function `e47::App::ProcessorWindow::timerCallback()':
App.cpp:(.text._ZN3e473App15ProcessorWindow13timerCallbackEv[_ZN3e473App15ProcessorWindow13timerCallbackEv]+0x13e): undefined reference to `e47::captureScreen(juce::Rectangle<int>)'
collect2: error: ld returned 1 exit status
make: *** [Makefile:109: build/AudioGridderServer] Error 1````

**Comment by apohl79** [comment]
Good to know thx. Linux client support might be interesting indeed. Regarding the server, I'm not sure if its worth the effort unfortunately, as most vendors support OS X and Windows only. Will keep this open.

**Comment by prene1** [comment]
Ive been trying to install and use this but im getting .dll errors

**Comment by apohl79** [comment]
Does this happen with both the latest and the alpha10-release build? If yes, can you post the exact errors?

**Comment by apohl79** [comment]
Ok figured it out. For now you have to install the following framework library manually: https://aka.ms/vs/16/release/VC_redist.x64.exe (Microsoft Visual C++ Redistributable für Visual Studio 2019, listed here: https://visualstudio.microsoft.com/de/downloads/).

I will package this into the setup in future builds.

**Comment by prene1** [comment]
> Ok figured it out. For now you have to install the following framework library manually: https://aka.ms/vs/16/release/VC_redist.x64.exe (Microsoft Visual C++ Redistributable für Visual Studio 2019, listed here: https://visualstudio.microsoft.com/de/downloads/).
> 
> I will package this into the setup in future builds.

Ok I will try. Btw I'm using windows 10 LTSC. 
Ill remote into my studio PC and give it a try.

**Comment by prene1** [comment]
> Ok figured it out. For now you have to install the following framework library manually: https://aka.ms/vs/16/release/VC_redist.x64.exe (Microsoft Visual C++ Redistributable für Visual Studio 2019, listed here: https://visualstudio.microsoft.com/de/downloads/).
> 
> I will package this into the setup in future builds.

Ok so on my normal window 10 pro production laptop. It worked fine. 

On my studio computer windows 10 LTSC I get,

VCRUNTIME140_1D.dll was not found. 
VCRUNTIME140D.dll " " 
MSVCP140.dll" " 
ucrtbased.dll " "

**Comment by elli-one** [comment]
This a very old problem for developper :)  Releae a DEBUG binary, build with the VC runtime DLL. For the developper PC with installed VS 2019 no problem, but for each other.  
Andreas you must creare a RELEASE binary. It is possible to build a release with more DEBUG options.
Or buid a DEBUG build with static VC runtime.

**Comment by apohl79** [comment]
Yeah static linking sounds like the best approach. I’m packaging release builds already. Thx! I’m really no windows expert! :-)

**Comment by rcrath** [comment]
> Good to know thx. Linux client support might be interesting indeed. Regarding the server, I'm not sure if its worth the effort unfortunately, as most vendors support OS X and Windows only. Will keep this open.

Actually on the server side rather than client, Linux is a lot more prevalent, so it might be worthwhile.

**Comment by KottV** [comment]
> Actually on the server side rather than client, Linux is a lot more prevalent, so it might be worthwhile.

Just curious, what kind of tasks Linux plugin server can do?

**Comment by rcrath** [comment]
@KottV (and of course @apohl79 ):

For an audiogridder Linux server, a docker container or a Snap might be a better solution than static libraries to make this more accessible on the Linux side...get it setup once as a completely self-contained unit with audio I/O and it could be insrted in the Jack Audio environment on any linux system with Jack in it.  You can set up web server containers either way, so the networking should be easy.  I could try doing this if I could get the server to build from source on a linux system, but I'm not an expert in containers or anything though I would be willing to try it out, as I have fumbled through the process of making them one time previously!
My use case: I have an aging windows laptop that groans under newer software, a linux computer that I do other work on but also have linux music stuff on, a good wired connection for both.  Many VSTs are now available on linux, so the idea would be to use this linux box or even an old unused computer setup just for this purpose (you can run linux, and linux audio at much lower latencies than win or Mac and it will run on pretty low spec, old hardware if you get the right linux distribution) , and pipe thus: Audio in ==> windows client audiogridder ==> linux audiogridder server == processed audio ==> win audiogridder client ==> audio out from windows client to rest of audio chain on win 10.  You can set up a linux box for audio with VST support in less than an hour with low latency if you have an old laptop or desktop lying around unused on the linux side.  For me, if it worked it would double the VST power of my old laptop!  

Other use cases: Harrison MixBus runs on linux and would make a great server host.  There are also a bunch of "headless" (no GUI) sound processing programs (ecasound and sox come to mind, but there may be more and I'm not sure about VST support there though I think they have it). There is a superb music ecosystem on the platform, and the whole raspberry pie and Arduino thing would be a great project to set up as a VST server (there are specific ones for audio...IdaFruit is one with low latency I think and good audio and midi I/O, but I'm not sure I am remembering right).  Since audiogridder uses TCP, non-real-time VST server could be set up on the web to grind through your heavy duty vst processing in a mixing situation rather than a playing one...there are already hardware versions of this where you rent time on someone's expensive compressor and stream audio through it, not in real time, and get back the processed audio.  Non-realtime is a pain because you cannot adjust on the fly, but still, If I only needed that $$$$-VST.com plugin for one track one time and wish I could do that, well there ya go. 

With Linux stuff you have a lot more room to experiment with things that might or might not work on the server side.  Windows and MacOS tend to lock things down more.  And on the server side in general (think internet, mobile devices, IoT) Linux is a much bigger player than on the client/app side because of all this flexibility, even though the learning curve is sometimes a pain..

**Comment by KottV** [comment]
As I can understand, main goal of Audiogridder is the plugin hosting, VST and AU.
Amount of Linux VST is lean in compare of Window or Mac. For routing audio and MIDI between devices you can use NetJack.

**Comment by rcrath** [comment]
You asked for use cases of audiogriddder as a Linux server and I gave them, so I am not sure what you are after, but I'll try to answer your concerns again more clearly.

TL;DR: 1) AudioGridder, if I understand right, is about moving audio across a network so that some plugins can reside on a different machine from the user's DAW/plugin host, not simply "plugin hosting."  2) Number of installs is not a very good metric of the value of a platform, but I respect Dev's decisions in that regard, 3. Netjack won't work in practice between Linux and Windows because it is buggy and crashes everything on the windows side. 

`For routing audio and MIDI between devices you can use NetJack.`

The use case I gave was serving network audio data between a Linux AudioGridder server and a Windows AudiGridder client.  Connecting Windows to Linux by NetJack involves installing Jack or Jack2 on Windows.  Jack and Jack2 on Windows crashes incessantly and renders ASIO useless to perform any tasks such as running a DAW to host VSTs in. It works in theory but not in practice.

`main goal of Audiogridder is the plugin hosting, VST and AU.`

Yes, in a client-server architecture where the client and server are on different computers connected over a network, which is quite a bit more than just hosting plugins, which any DAW can do on any platform.  The use case I gave above is to serve VST plugins from a Linux computer to extend the capabilities of the Windows computer, just as you would from a Mac or Win computer, but with the advantage of the ability to run on old unused computers or new cheap single purpose computers, or even a supercomputing cluster if you have access.  Because it is open source, Linux is much more flexible than Win or Mac and opens up a bunch of possibilities beyond what can be done on those platforms.  For example, If I understand right, once the Linux AGddr server receives the audio and sends it to a vst, that VST could probably send the audio out to either another VST or anything else in the Linux audio domain, which opens up a great number of audio-treating possibilities not available on Win or Mac.  

I understand the economic argument coming from number of installs, so no need to rehash that here. Devs have limited time and I respect that. Especially in the open source world. That is why I offered to help rather than just offer opinions. 

Linux as a server offers possibilities for innovation and economy that Win and Mac cannot.  I think leaving it out, especially when developing with the JUCE framework which can compile all three OSs and more, is to miss out on a really useful application of the server part of the server/client Audiogridder pair.  The client is already available on Linux, so I am asking the dev(s) to keep the ability to compile the server portion on Linux available and not to shut it down just yet, which @apohl79 indicated elsewhere he was thinking about.

**Comment by KottV** [comment]
Ok, i didn't test NetJack on Windows, so maybe it really useless on that platform. 

> For example, If I understand right, once the Linux AGddr server receives the audio and sends it to a vst, that VST could probably send the audio out to either another VST or anything else in the Linux audio domain

This audio is sent back to AudioGridder host (with video) and then back to client. Audio/MIDI and video of the plugin's interface, nothing more. So I still don't see what  AudioGridder server can do on Linux, just few Linux VSTs on old computers or just couples of Linux ARM VSTs on RPi?

What you described - looks like another jack/rewire technique. So I think it's better to improve Jack and JackAss https://github.com/falkTX/JackAss for such tasks.

**Comment by apohl79** [comment]
I think - taking the use cases discussion aside - it should be fairly easy to get the server to run on linux with its current code base. No screen capturing or mouse/keyboard support though, as this has to be ported to X. But using the generic editor of the AG plugin, it should be possible to control remote VSTs. 

I would expect a few minor code changes to be necessary, basically to deactivate OSX code not working on linux. Could be worth a try.

**Comment by rcrath** [comment]
@KottV There is a communication problem here that is making things difficult since you have repeatedly mischaracterized what I have said and sent me for "better" solutions that don't apply.  I got that you think an AG server on linux is a bad useless idea from your first question and subsequent responses. I also understand what AG does, which is why I'm interested in this project.  Let's leave it at that, please.  

Thanks @apohl79 for being open to keeping the server code going. I'll see if I can get things going on my setup and report back.

**Comment by DCTech-lab** [comment]
> @KottV (and of course @apohl79 ):
> 
> For an audiogridder Linux server, a docker container or a Snap might be a better solution than static libraries to make this more accessible on the Linux side...get it setup once as a completely self-contained unit with audio I/O and it could be insrted in the Jack Audio environment on any linux system with Jack in it. You can set up web server containers either way, so the networking should be easy. I could try doing this if I could get the server to build from source on a linux system, but I'm not an expert in containers or anything though I would be willing to try it out, as I have fumbled through the process of making them one time previously!
> My use case: I have an aging windows laptop that groans under newer software, a linux computer that I do other work on but also have linux music stuff on, a good wired connection for both. Many VSTs are now available on linux, so the idea would be to use this linux box or even an old unused computer setup just for this purpose (you can run linux, and linux audio at much lower latencies than win or Mac and it will run on pretty low spec, old hardware if you get the right linux distribution) , and pipe thus: Audio in ==> windows client audiogridder ==> linux audiogridder server == processed audio ==> win audiogridder client ==> audio out from windows client to rest of audio chain on win 10. You can set up a linux box for audio with VST support in less than an hour with low latency if you have an old laptop or desktop lying around unused on the linux side. For me, if it worked it would double the VST power of my old laptop!
> 
> Other use cases: Harrison MixBus runs on linux and would make a great server host. There are also a bunch of "headless" (no GUI) sound processing programs (ecasound and sox come to mind, but there may be more and I'm not sure about VST support there though I think they have it). There is a superb music ecosystem on the platform, and the whole raspberry pie and Arduino thing would be a great project to set up as a VST server (there are specific ones for audio...IdaFruit is one with low latency I think and good audio and midi I/O, but I'm not sure I am remembering right). Since audiogridder uses TCP, non-real-time VST server could be set up on the web to grind through your heavy duty vst processing in a mixing situation rather than a playing one...there are already hardware versions of this where you rent time on someone's expensive compressor and stream audio through it, not in real time, and get back the processed audio. Non-realtime is a pain because you cannot adjust on the fly, but still, If I only needed that $$$$-VST.com plugin for one track one time and wish I could do that, well there ya go.
> 
> With Linux stuff you have a lot more room to experiment with things that might or might not work on the server side. Windows and MacOS tend to lock things down more. And on the server side in general (think internet, mobile devices, IoT) Linux is a much bigger player than on the client/app side because of all this flexibility, even though the learning curve is sometimes a pain..

Waves Soundgrid server on linux (their soundgrid plugins have a linux binary in bundle)
